src/circuit/blinding_circuit.rs

```diff
@@ -1,173 +1,102 @@
-    pub verifier: Verifier<CP::CurveBaseField, CP::Curve, CP::OuterCurvePC>,
-    pub vk: <CP::OuterCurvePC as PolynomialCommitment<
-        CP::CurveBaseField,
-        DensePolynomial<CP::CurveBaseField>,
-    >>::VerifierKey,
+    pub vk: VerifierKey<CP::CurveScalarField, CP::CurvePC>,
+    pub blinding: Blinding<CP::CurveScalarField>,
+    pub zh: [CP::CurveBaseField; 2],
+    pub private_inputs: Vec<CP::CurveBaseField>,
 }
 
-impl<CP: CircuitParameters> BlindingCircuit<CP> {
-    pub fn precompute_prover(
-        setup: &<CP::OuterCurvePC as PolynomialCommitment<
-            CP::CurveBaseField,
-            DensePolynomial<CP::CurveBaseField>,
-        >>::UniversalParams,
-        gadget: fn(
-            &mut StandardComposer<CP::CurveBaseField, CP::Curve>,
-            private_inputs: &[CP::CurveBaseField],
-            public_inputs: &[CP::CurveBaseField],
-        ) -> Vec<Point<CP::Curve>>,
-        private_inputs: &[CP::CurveBaseField],
-        public_inputs: &[CP::CurveBaseField],
-    ) -> (
-        // Prover
-        Prover<CP::CurveBaseField, CP::Curve, CP::OuterCurvePC>,
-        // CommitterKey
-        <CP::OuterCurvePC as PolynomialCommitment<
-            CP::CurveBaseField,
-            DensePolynomial<CP::CurveBaseField>,
-        >>::CommitterKey,
-        // VerifierKey
-        <CP::OuterCurvePC as PolynomialCommitment<
-            CP::CurveBaseField,
-            DensePolynomial<CP::CurveBaseField>,
-        >>::VerifierKey,
-        // PublicInput
-        PublicInputs<CP::CurveBaseField>,
-    ) {
-        // Create a `Prover`
-        // Set the circuit using `gadget`
-        // Output `prover`, `vk`, `public_input`.
-
-        let mut prover = Prover::<CP::CurveBaseField, CP::Curve, CP::OuterCurvePC>::new(b"demo");
-        prover.key_transcript(b"key", b"additional seed information");
-        gadget(prover.mut_cs(), private_inputs, public_inputs);
-        let (ck, vk) = CP::OuterCurvePC::trim(
-            setup,
-            prover.circuit_bound().next_power_of_two() + 6,
-            0,
-            None,
-        )
-        .unwrap();
-        let public_input = prover.mut_cs().get_pi().clone();
-
-        (prover, ck, vk, public_input)
-    }
+impl<CP> Circuit<CP::CurveBaseField, CP::Curve> for BlindingCircuit<CP>
+where
+    CP: CircuitParameters,
+{
+    const CIRCUIT_ID: [u8; 32] = [0x01; 32];
+
+    // Default implementation
+    fn gadget(
+        &mut self,
+        composer: &mut StandardComposer<CP::CurveBaseField, CP::Curve>,
+    ) -> Result<(), Error> {
+        // parse the public inputs (todo is Com(Z_H) a public input?)
+        let com_z_h = TEGroupAffine::<CP::Curve>::new(self.zh[0], self.zh[1]);
+
+        assert_eq!(self.private_inputs.len() % 3, 0);
+        let mut i = 0;
+        let mut q: Point<CP::Curve>;
+        while i < self.private_inputs.len() {
+            // parse the private inputs
+            q = composer.add_affine(TEGroupAffine::<CP::Curve>::new(
```

Comment:
We need to double check if 2 random scalars are enough in terms of security (even if this will probably be replaced later).